MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc

```diff
@@ -0,0 +1,207 @@
+  //   Global(for vol at pi/2)    Local
+  //   +R (+Y)                    +Z
+  //   +phi(-X)                   +X
+  //   +Z                         +Y
+  //
+  // Note that the frame is centered in the DDD volume center, which is
+  // inside the volume for DDD boxes and (pesudo)trapezoids, on the beam line
+  // for tubs, cons and trunctubs.
+
+  // In geometry version 1103l, trapezoids have X and Z in the opposite direction
+  // than the above.  Boxes are either oriented as described above or in some case
+  // have opposite direction for Y and X.
+
+  // The global position
+  Surface::PositionType &posResult = center_;
+
+  // The reference plane rotation
+  math::XYZVector x, y, z;
+  dd4hep::Rotation3D refRot;
+  fv.rot(refRot);
+  refRot.GetComponents(x, y, z);
+  if (debug) {
+    if (x.Cross(y).Dot(z) < 0.5) {
+      LogDebug("magneticfield::volumeHandle") << "*** WARNING: Rotation is not RH ";
+    }
+  }
+
+  // The global rotation
+  Surface::RotationType rotResult(float(x.X()),
+                                  float(x.Y()),
+                                  float(x.Z()),
+                                  float(y.X()),
+                                  float(y.Y()),
+                                  float(y.Z()),
+                                  float(z.X()),
+                                  float(z.Y()),
+                                  float(z.Z()));
+
+  refPlane = new GloballyPositioned<float>(posResult, rotResult);
+
+  // Check correct orientation
+  if (debug) {
+    LogDebug("magneticfield::volumeHandle") << "Refplane pos  " << refPlane->position();
+
+    // See comments above for the conventions for orientation.
+    LocalVector globalZdir(0., 0., 1.);  // Local direction of the axis along global Z
+
+    /* Preserve in case pseudotrap is needed again
+    if (shape_ == DDSolidShape::ddpseudotrap) {
+      globalZdir = LocalVector(0., 1., 0.);
+    }
+    */
+    if (refPlane->toGlobal(globalZdir).z() < 0.) {
+      globalZdir = -globalZdir;
+    }
+    float chk = refPlane->toGlobal(globalZdir).dot(GlobalVector(0, 0, 1));
+    if (chk < .999)
+      LogDebug("magneticfield::volumeHandle") << "*** WARNING RefPlane check failed!***" << chk;
+  }
+}
+
+std::vector<VolumeSide> volumeHandle::sides() const {
+  std::vector<VolumeSide> result;
+  for (int i = 0; i < 6; ++i) {
+    // If this is just a master volume out of wich a 2pi volume
+    // should be built (e.g. central cylinder), skip the phi boundaries.
+    if (expand && (i == phiplus || i == phiminus))
+      continue;
+
+    // FIXME: Skip null inner degenerate cylindrical surface
+    if (shape_ == DDSolidShape::ddtubs && i == SurfaceOrientation::inner && theRMin < 0.001)
+      continue;
+
+    ReferenceCountingPointer<Surface> s = const_cast<Surface *>(surfaces[i].get());
+    result.push_back(VolumeSide(s, GlobalFace(i), surfaces[i]->side(center_, 0.3)));
+  }
+  return result;
+}
+
+// To allow the following code to work with both old and new DD,
```

Comment:
@ianna @cvuosalo I'm not sure of what you mean with "work with both old and new DD". I understood that as of now this code is run to parse a given xml with the new DD implementation and the original one to parse the same xml with the old DD implementation. So I am a bit confused on what it means that this  code will work on both?

Reply:
@namapane: The files included below, `buildBox.icc` and the rest, are included in both the old DD version of `volumeHandle.cc` and this DD4hep version of `volumeHandle.cc`. Currently, the old DD returns lengths as millimeters, which is the Geant4 convention, while DD4hep returns lengths as centimeters, which is the CMS convention. `buildBox.icc`, for example, queries the DD to get the lengths of the sides of a box. To get the right length conversion using the same source code, I define the `convertUnits` function just below differently for the old DD and for DD4hep.

Note that if DD4hep is changed to return lengths following the Gean4 convention, as you and Yana have discussed, then this problem goes away, and the same unit conversion (mm->cm) would be used for both old and new.

Reply:
I see, the comment refers to the common part only (icc files). I thought that it referred to the "updated" volumeHandle.cc as a whole.

Reply:
@namapane: I clarified the comment. I will wait to push the change after Slava finishes reviewing this PR in case other changes need to be made.

Reply:
> @namapane: I clarified the comment. I will wait to push the change after Slava finishes reviewing this PR in case other changes need to be made.

I was not planning to review this in detail until @namapane is happy with the code.